### ISSUE TITLE
Calculating NDCG for a Keras model during run-time.

### ISSUE THREAD

**Issue by anktplwl91** [github/CodeSearchNet#59]
Hi

I am trying to build a Keras model from scratch using only Python dataset for a start. But I am confused as to how NDCG can be calculated during training or even testing.

According to my understanding, calculating NDCG will require a ground truth of which pages are relevant to a query in a ranked order and the model's predictions for that query. But in the dataset, each code-block is provided with its own query (docstring), but no ranking order of pages for a query.

I am using same network architecture as proposed in this repository, but only have python as the language input for now, the final layer being a 2D matrix (softmax applied) of shape (BS*BS), with each cell having relevance score for (query, code-page) pair. As ground-truth to this, I prepared a 2D matrix having 1 as diagonal elements and 0 elsewhere.

But how can I calculate NDCG in this scenario, as only 1 page is relevant for each query, instead of a relevance list of pages?

**Comment by mallamanis** [comment]
Hi @anktplwl91 the human-annotated relevance data are the ones used for computing NDCG. There is no data in this repo that would allow you to compute an NDCG. Once you submit to the leaderboard, the NDCG on the test data will be computed.

For training/validation, once can compute some proxy metric. We use MRR for validation (note that you cannot directly differentiate through MRR) and a classification-like loss for training. Have a look at the [technical report](https://arxiv.org/abs/1909.09436) for more details.
